components/brave_wallet/browser/json_rpc_service.cc

```diff
@@ -2120,21 +2120,73 @@ void JsonRpcService::GetERC1155TokenBalance(
-  auto callback = base::BindOnce(&JsonRpcService::OnDiscoverAssetsCompleted,
-                                 weak_ptr_factory_.GetWeakPtr());
-  DiscoverAssetsInternal(chain_id, coin, account_addresses,
-                         std::move(callback));
+  // Simple client side rate limiting (only applies to refreshes)
+  const base::Time assets_last_discovered_at =
+      prefs_->GetTime(kBraveWalletLastDiscoveredAssetsAt);
+  if (!assets_last_discovered_at.is_null() &&
+      ((base::Time::Now() - base::Minutes(kAssetDiscoveryMinutesPerRequest)) <
+       assets_last_discovered_at)) {
+    return;
+  }
+  prefs_->SetTime(kBraveWalletLastDiscoveredAssetsAt, base::Time::Now());
+
+  // Fetch block numbers for which asset discovery has been run through
+  auto next_asset_discovery_from_blocks =
+      prefs_->GetDict(kBraveWalletNextAssetDiscoveryFromBlocks).Clone();
+  for (std::string chain_id : GetAssetDiscoverySupportedChains()) {
+    auto internal_callback =
+        base::BindOnce(&JsonRpcService::OnDiscoverAssetsCompleted,
+                       weak_ptr_factory_.GetWeakPtr());
+
+    // Call DiscoverAssets for the supported chain ID
+    // using the kBraveWalletNextAssetDiscoveryFromBlocks pref
+    // as the from_block of the eth_getLogs query.
+    std::string from_block = "earliest";
+    std::string to_block = "latest";
+    const std::string* next_asset_discovery_from_block =
+        next_asset_discovery_from_blocks.FindString(chain_id);
+    if (next_asset_discovery_from_block) {
+      from_block = *next_asset_discovery_from_block;
+    }
+
+    JsonRpcService::DiscoverAssets(chain_id, mojom::CoinType::ETH,
+                                   account_addresses, true, from_block,
+                                   to_block, std::move(internal_callback));
+  }
+}
+
+void JsonRpcService::DiscoverAssetsOnAllSupportedChains(
+    const std::vector<std::string>& account_addresses) {
+  for (std::string chain_id : GetAssetDiscoverySupportedChains()) {
+    auto internal_callback =
+        base::BindOnce(&JsonRpcService::OnDiscoverAssetsCompleted,
+                       weak_ptr_factory_.GetWeakPtr());
+    JsonRpcService::DiscoverAssets(chain_id, mojom::CoinType::ETH,
+                                   account_addresses, false, "earliest",
+                                   "latest", std::move(internal_callback));
+  }
+}
+
+// Call observers
+void JsonRpcService::CompleteDiscoverAssets(
+    const std::string& chain_id,
+    std::vector<mojom::BlockchainTokenPtr> discovered_assets,
+    mojom::ProviderError error,
+    const std::string& error_message,
+    DiscoverAssetsCallback callback) {
+  for (const auto& observer : observers_) {
+    observer->OnDiscoveredAssetsCompleted(chain_id, error, error_message);
+  }
+
+  std::move(callback).Run(chain_id, std::move(discovered_assets), error,
+                          error_message);
 }
 
 void JsonRpcService::OnDiscoverAssetsCompleted(
```

Comment:
Is this callback really useful atm?
I don't quite see the need and we can output error log in CompleteDiscoverAssets if we really want it.

Reply:
This is just useful for testing so that I can pass a lambda that runs assertions on the expected chain_id, discovered_assets, error, and error messages.

Reply:
Updated here https://github.com/brave/brave-core/pull/15707/commits/ea8d7de38f0274da10aed2242c2449f315bd673c